test/parallel_api/numeric/numeric.ops/reduce_by_segment.pass.cpp

```diff
@@ -21,210 +21,219 @@
+
+        auto new_policy2 = make_new_policy<new_kernel_name<Policy, 1>>(exec);
+        auto res2 = oneapi::dpl::reduce_by_segment(new_policy2, keys_first, keys_last, vals_first, key_res_first,
+                                                   val_res_first, ::std::equal_to<KeyT>());
+#    if _PSTL_SYCL_TEST_USM
+        exec.queue().wait_and_throw();
+#    endif
+        result_size = std::distance(key_res_first, res2.first);
+
+        {
+            auto host_key_res = get_host_pointer(key_res_first);
+            auto host_val_res = get_host_pointer(val_res_first);
+            check_values(host_key_res, host_val_res, result_size);
         }
-    }
 
-    // call algorithm on single element range
-    key_res_head[0] = 9;
-    val_res_head[0] = 9;
+        // call algorithm with addition operator
+        {
+            auto host_keys = get_host_pointer(keys_first);
+            auto host_vals = get_host_pointer(vals_first);
+            auto host_key_res = get_host_pointer(key_res_first);
+            auto host_val_res = get_host_pointer(val_res_first);
 
-    auto new_policy2 = oneapi::dpl::execution::make_device_policy<class reduce_by_segment_2>(q);
-    auto res2 = oneapi::dpl::reduce_by_segment(new_policy2, key_head, key_head + 1, val_head, key_res_head, val_res_head);
+            initialize_data(host_keys, host_vals, host_key_res, host_val_res, n);
+        }
 
-    // check values
-    n = std::distance(key_res_head, res2.first);
-    ASSERT_EQUAL(n, 1);
-    ASSERT_EQUAL(key_res_head[0], 1);
-    ASSERT_EQUAL(val_res_head[0], 1);
+        auto new_policy3 = make_new_policy<new_kernel_name<Policy, 2>>(exec);
+        auto res3 = oneapi::dpl::reduce_by_segment(new_policy3, keys_first, keys_last, vals_first, key_res_first,
+                                                   val_res_first, ::std::equal_to<KeyT>(), ::std::plus<ValT>());
+#    if _PSTL_SYCL_TEST_USM
+        exec.queue().wait_and_throw();
+#    endif
+        result_size = std::distance(key_res_first, res3.first);
+
+        auto host_key_res = get_host_pointer(key_res_first);
+        auto host_val_res = get_host_pointer(val_res_first);
+        check_values(host_key_res, host_val_res, result_size);
+    }
+#endif
 
-    // Deallocate memory
-    sycl::free(key_head, q);
-    sycl::free(val_head, q);
-    sycl::free(key_res_head, q);
-    sycl::free(val_res_head, q);
-}
+    // specialization for host execution policies
+    template <typename Policy, typename Iterator1, typename Iterator2, typename Iterator3, typename Iterator4,
+              typename Size>
+    typename ::std::enable_if<
+#if TEST_DPCPP_BACKEND_PRESENT
+        !oneapi::dpl::__internal::__is_hetero_execution_policy<typename ::std::decay<Policy>::type>::value &&
 #endif
+            !is_same_iterator_category<Iterator3, ::std::bidirectional_iterator_tag>::value &&
+            !is_same_iterator_category<Iterator3, ::std::forward_iterator_tag>::value,
```

Comment:
The specification defines such API:
```c++
template<typename Policy, typename InputKeyIt, typename InputValueIt,
    typename OutputKeyIt, typename OutputValueIt,
    typename BinaryPredcate =
        std::equal_to<typename std::iterator_traits<InputKeyIt>::value_type,
    typename BinaryOp =
        std::plus<typename std::iterator_traits<InputValueIt>::value_type>>
std::pair<OutputKeyIt,OutputValueIt>
reduce_by_segment(Policy&& policy, InputKeyIt keys_first, InputKeyIt keys_last,
    InputValueIt values_first, OutputKeyIt keys_result,
    OutputValueIt values_result,
    BinaryPred binary_pred =
        std::equal_to<typename std::iterator_traits<InputKeyIt>::value_type>(),
    BinaryOp binary_op =
        std::plus<typename std::iterator_traits<InputValueIt>::value_type>());
```
As I understood, the algorithm needs input iterator for input sequences and output iterator for resulting sequences. Should we test forward and bidirectional iterators as well?